[WIP] SkipSkan

JLockerman opened the pull request:
Best review order is probably:

1. the README, which explains what we're trying to do
2. exec.c, which contains the code doing it, and some more detailed docs
3. planner.c, which contains the heuristics we use to detect when a SkipSkan is viable

JLockerman committed:
Implement SkipSkan tp speed SELECT DISTINCT

davidkohn88 commented on tsl/src/nodes/skip_skan/exec.c:
```diff
@@ -0,0 +1,396 @@
+	{
+		IndexScanState *idx = castNode(IndexScanState, ExecInitNode(state->idx_scan, estate, eflags));
+
+		node->custom_ps = list_make1(&idx->ss.ps);
```
Why do we initialize our plan state to the indexscan's plan state?

davidkohn88 commented on tsl/src/nodes/skip_skan/exec.c:
```diff
@@ -0,0 +1,396 @@
+	{
+		/* first time through we ignore the inital scan keys, which are used to
+		 * skip previously seen values, we'll change back the number of scan keys
+		 * the first time through update_skip_key
```
Revise comment here.

davidkohn88 reviewed (commented):
Some comments from our session.

JLockerman committed:
start refactor for single column

JLockerman committed:
variable renames

JLockerman committed:
Enable SkipSkan with RuntimeKeys

The previous disable was based on a misread of the IndexScan code, which
missed the RuntimeKeys array.

JLockerman committed:
The great refactor to an explicit statemachine

JLockerman committed:
add state machine comment

JLockerman committed:
Remove unneeded headers from skip_skan/exec.c

JLockerman committed:
Refactor SkipSkan planner for single keys

JLockerman committed:
Enable SkipSkans on columns other than the first column of the index.
Notably, this allows us to emulate multi-column SkipScan using LATERAL

JLockerman committed:
Add more tests for SkipSkans on non-first columns
